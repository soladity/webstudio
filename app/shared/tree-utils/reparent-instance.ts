import produce from "immer";
import { type Instance } from "@webstudio-is/sdk";
import { findInstanceById } from "./find-instance";
import { deleteInstanceMutable } from "./delete-instance";
import {
  insertInstanceMutable,
  type InstanceInsertionSpec,
} from "./insert-instance";

export type InstanceReparentingSpec = {
  id: Instance["id"];
  parentId: Instance["id"];
  position: InstanceInsertionSpec["position"];
};

export const reparentInstance = produce(
  (draftRootInstance: Instance, spec: InstanceReparentingSpec) => {
    const instance = findInstanceById(draftRootInstance, spec.id);
    if (instance === undefined) return;
    deleteInstanceMutable(draftRootInstance, instance.id);
    insertInstanceMutable(draftRootInstance, instance, spec);
  }
);
