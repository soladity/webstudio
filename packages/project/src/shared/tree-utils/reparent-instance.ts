import type { Instance } from "@webstudio-is/project-build";
import { findInstanceById } from "./find-instance";
import { findParentInstance } from "./find-parent-instance";
import { deleteInstanceMutable } from "./delete-instance";
import { insertInstanceMutable } from "./insert-instance";

export const reparentInstanceMutable = (
  rootInstance: Instance,
  instanceId: Instance["id"],
  newParentId: Instance["id"],
  newPosition: number | "end"
) => {
  const instance = findInstanceById(rootInstance, instanceId);
  if (instance === undefined) {
    return;
  }

  if (newPosition === "end") {
    deleteInstanceMutable(rootInstance, instance.id);
    insertInstanceMutable(rootInstance, instance, {
      parentId: newParentId,
      position: "end",
    });
    return;
  }

  let newPositionAdjusted = newPosition;

  // If parent is the same, we need to adjust the position to account for the removal of the instance.
  const currentParent = findParentInstance(rootInstance, instance.id);
  if (currentParent !== undefined && currentParent.id === newParentId) {
    const currentPosition = currentParent.children.findIndex(
      (child) => child.type === "instance" && child.id === instance.id
    );
    if (currentPosition < newPosition) {
      newPositionAdjusted--;
    }
  }

  deleteInstanceMutable(rootInstance, instance.id);
  insertInstanceMutable(rootInstance, instance, {
    parentId: newParentId,
    position: newPositionAdjusted,
  });
};
