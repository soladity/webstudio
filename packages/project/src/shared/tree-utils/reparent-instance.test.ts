import { describe, test, expect } from "@jest/globals";
import type { Instance } from "@webstudio-is/project-build";
import { reparentInstanceMutable } from "./reparent-instance";

describe("Reparent instance", () => {
  test(`new parent`, () => {
    const rootInstance: Instance = {
      type: "instance",
      component: "Box",
      id: "1",
      children: [
        {
          type: "instance",
          component: "Box",
          id: "2",
          children: [],
        },
        {
          type: "instance",
          component: "Box",
          id: "3",
          children: [
            {
              type: "instance",
              component: "Box",
              id: "4",
              children: [],
            },
          ],
        },
      ],
    };

    reparentInstanceMutable(rootInstance, "4", "2", 0);
    expect(rootInstance).toMatchSnapshot();
  });

  test(`same parent different position`, () => {
    const rootInstance: Instance = {
      type: "instance",
      component: "Box",
      id: "1",
      children: [
        {
          type: "instance",
          component: "Box",
          id: "2",
          children: [],
        },
        {
          type: "instance",
          component: "Box",
          id: "3",
          children: [],
        },
        {
          type: "instance",
          component: "Box",
          id: "4",
          children: [],
        },
      ],
    };

    reparentInstanceMutable(rootInstance, "2", "1", 2);
    expect(rootInstance).toMatchSnapshot();
  });
});
