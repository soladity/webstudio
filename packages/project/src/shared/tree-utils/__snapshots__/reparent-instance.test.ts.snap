// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reparent instance new parent 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "component": "Box",
          "cssRules": [],
          "id": "4",
        },
      ],
      "component": "Box",
      "cssRules": [],
      "id": "2",
    },
    {
      "children": [],
      "component": "Box",
      "cssRules": [],
      "id": "3",
    },
  ],
  "component": "Box",
  "cssRules": [],
  "id": "1",
}
`;

exports[`Reparent instance same parent different position 1`] = `
{
  "children": [
    {
      "children": [],
      "component": "Box",
      "cssRules": [],
      "id": "3",
    },
    {
      "children": [],
      "component": "Box",
      "cssRules": [],
      "id": "2",
    },
    {
      "children": [],
      "component": "Box",
      "cssRules": [],
      "id": "4",
    },
  ],
  "component": "Box",
  "cssRules": [],
  "id": "1",
}
`;
